addon/lib/webextension-channels.js

```diff
@@ -0,0 +1,147 @@
+/*
+ * This Source Code is subject to the terms of the Mozilla Public License
+ * version 2.0 (the 'License'). You can obtain a copy of the License at
+ * http://mozilla.org/MPL/2.0/.
+ */
+
+/* global XPCOMUtils, Services */
+
+const { Ci, Cu } = require('chrome');
+const { Class } = require('sdk/core/heritage');
+const { Disposable } = require('sdk/core/disposable');
+
+const store = require('sdk/simple-storage').storage;
+
+const Metrics = require('./metrics');
+
+const TESTPILOT_TELEMETRY_CHANNEL = 'testpilot-telemetry';
+
+Cu.import('resource://gre/modules/XPCOMUtils.jsm');
+
+// Lazy imports.
+XPCOMUtils.defineLazyModuleGetter(this, 'Services',
+                                  'resource://gre/modules/Services.jsm');
+
+const {getExtensionUUID} = Cu.import('resource://gre/modules/Extension.jsm', {});
+
+function createChannelForAddonId(name, addonId) {
+  const targetExtensionUUID = getExtensionUUID(addonId);
+  const baseURI = Services.io
+        .newURI(`moz-extension://${targetExtensionUUID}/_blank.html`, null, null);
+  const principal = Services.scriptSecurityManager
+        .createCodebasePrincipal(baseURI, { addonId });
+
+  const addonChromeWebNav = Services.appShell.createWindowlessBrowser(true);
+  const docShell = addonChromeWebNav.QueryInterface(Ci.nsIInterfaceRequestor) // eslint-disable-line new-cap
+        .getInterface(Ci.nsIDocShell);
+  docShell.createAboutBlankContentViewer(principal);
+  const window = docShell.contentViewer.DOMDocument.defaultView;
```

Comment:
> Honestly, I didn't really know you could get a DOM window pointer for another origin from add-on code :)

@lmorchard Ah yeah, once you `require('chrome')` you effectively have rootly powers, and a lot of weird stuff becomes possible ^_^

After looking at this snippet for a while, and reading through https://bugzil.la/1186732, I'm much less scared of it. But it might be nice to document it for the sake of our future selves. Here's an attempt, take whatever seems useful, or skip it if it seems to make things more confusing overall:

```js
function createChannelForAddonId(name, addonId) {
  // The BroadcastChannel API allows messaging between different windows that
  // share the same origin. Bug 1186732 extended this to WebExtensions (which
  // may not have an origin) by adding a special URL that loads an about:blank
  // page at the (generalized) "origin" of the extension.
  //
  // Load that about:blank page, and use its `window` to get a BroadcastChannel
  // that allows two-way communication between the main Test Pilot extension and
  // individual experiment extensions. 
  
  // Note: the `targetExtensionUUID` is different for each copy of Firefox,
  // so that malicious websites can't guess the special URL associated with
  // an extension. 
  const targetExtensionUUID = getExtensionUUID(addonId);
  
  // Create the special about:blank URL for the extension.
  const baseURI = Services.io
        .newURI(`moz-extension://${targetExtensionUUID}/_blank.html`, null, null);
  
  // Create a principal (security context) for the generalized origin given
  // by the extension's special URL and its `addonId`.
  const principal = Services.scriptSecurityManager
        .createCodebasePrincipal(baseURI, { addonId });
  
  // Create a hidden window and open the special about:blank page for the
  // extension.
  const addonChromeWebNav = Services.appShell.createWindowlessBrowser(true);
  const docShell = addonChromeWebNav.QueryInterface(Ci.nsIInterfaceRequestor) // eslint-disable-line new-cap
        .getInterface(Ci.nsIDocShell);
  docShell.createAboutBlankContentViewer(principal);
  const window = docShell.contentViewer.DOMDocument.defaultView;
  
  // Finally, get the BroadcastChannel associated with the extension.
  const addonBroadcastChannel = new window.BroadcastChannel(name);
  
  // Callers need to keep the pointer to the window, otherwise the window's 
  // BroadcastChannel will get garbage collected.
  return {
    addonChromeWebNav,
    addonBroadcastChannel
  };
} 
```